File: leetcode/find-minimum-in-rotated-sorted-array.java
//https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/

class Solution {
    public int findMin(int[] nums) {
        int hi = nums.length-1;
        int lo = 0;
        int mid = (hi+lo)/2;
        
        //array is not rotated
        if(nums[lo] < nums[hi]) return nums[lo];

        while(lo < hi){
            if(mid>0 && nums[mid] < nums[mid-1]) return nums[mid];
            
            if(nums[mid] < nums[hi]) hi = mid-1;
            else lo = mid+1;
            
            mid = lo+((hi-lo)/2);
        }
        
        return nums[lo];
    }
}

File: leetcode/find_minimum_in_rotated_sorted_array.py
# https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/description/


class Solution:
    def findMin(self, nums: list[int]) -> int:
        left: int = 0
        right: int = len(nums) - 1

        if len(nums) < 3:
            return min(nums)

        while left < right:
            mid: int = int((left + right) / 2)
            if mid > 0 and nums[mid] < nums[mid - 1]:
                return nums[mid]
            if nums[mid] > nums[left]:
                if nums[right] > nums[left]:
                    right = mid - 1
                else:
                    left = mid + 1
            elif nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid - 1
        return nums[left]
